### ISSUE TITLE
`incus_instance` fails on `image` with a public remote

### ISSUE THREAD

**Issue by benoitjpnet** [lxc/terraform-provider-incus#119]
Hello,

I have a public remote:

```
root@incus:~# incus remote ls
+-----------------+------------------------------------+---------------+-------------+--------+--------+--------+
|      NAME       |                URL                 |   PROTOCOL    |  AUTH TYPE  | PUBLIC | STATIC | GLOBAL |
+-----------------+------------------------------------+---------------+-------------+--------+--------+--------+
| 10.10.10.174    | https://10.10.10.174:8443          | incus         | tls         | YES    | NO     | NO     |
+-----------------+------------------------------------+---------------+-------------+--------+--------+--------+
| images          | https://images.linuxcontainers.org | simplestreams | none        | YES    | NO     | NO     |
+-----------------+------------------------------------+---------------+-------------+--------+--------+--------+
| local (current) | unix://                            | incus         | file access | NO     | YES    | NO     |
+-----------------+------------------------------------+---------------+-------------+--------+--------+--------+
root@incus:~#
```

```
root@incus:~# incus image ls 10.10.10.174:
+----------------------------+--------------+--------+-----------------------------------------------------+--------------+-----------+-----------+----------------------+
|           ALIAS            | FINGERPRINT  | PUBLIC |                     DESCRIPTION                     | ARCHITECTURE |   TYPE    |   SIZE    |     UPLOAD DATE      |
+----------------------------+--------------+--------+-----------------------------------------------------+--------------+-----------+-----------+----------------------+
| forgejo-8.0.1-1benoitjpnet | 37218922c92c | yes    | forgejo v8.0.1 on ubuntu/24.04 release 1benoitjpnet | x86_64       | CONTAINER | 234.06MiB | 2024/08/15 11:43 UTC |
+----------------------------+--------------+--------+-----------------------------------------------------+--------------+-----------+-----------+----------------------+
root@incus:~#
```

Tofu excerpt using this remote:

```
resource "incus_instance" "forgejo" {
  name  = "forgejo"
  image = "10.10.10.174:forgejo-8.0.1-1benoitjpnet"

  device {
  [...]
  }
}
```

It fails with:

```
incus_instance.forgejo: Creating...
╷
│ Error: Failed to retrieve Incus ImageServer
│
│   with incus_instance.forgejo,
│   on forgejo.tf line 6, in resource "incus_instance" "forgejo":
│    6: resource "incus_instance" "forgejo" {
│
│ The remote isn't a private server
╵
```

I wonder if this is a bug, or if I am missing something?
Thank you!

**Comment by benoitjpnet** [comment]
Also the doc at https://github.com/lxc/terraform-provider-incus/blob/main/docs/resources/instance.md contains a broken link:

```
* `image` - *Optional* - Base image from which the instance will be created. Must
  specify [an image accessible from the provider remote](https://linuxcontainers.org/incus/docs/main/reference/remote_image_servers/). << broken
```

**Comment by benoitjpnet** [comment]
TF_LOG=DEBUG:

```
incus_instance.forgejo: Creating...
2024-08-20T20:32:15.997+0900 [INFO]  Starting apply for incus_instance.forgejo
2024-08-20T20:32:15.998+0900 [DEBUG] skipping FixUpBlockAttrs
2024-08-20T20:32:15.998+0900 [DEBUG] incus_instance.forgejo: applying the planned Create change
2024-08-20T20:32:16.070+0900 [ERROR] provider.terraform-provider-incus_v0.1.3: Response contains error diagnostic: @caller=github.com/hashicorp/terraform-plugin-go@v0.23.0/tfprotov6/internal/diag/diagnostics.go:58 diagnostic_detail="The remote isn't a private server" tf_provider_addr=registry.terraform.io/lxc/incus tf_req_id=c1d00871-b6c0-caa6-46d9-ef8c2db61dfa tf_resource_type=incus_instance @module=sdk.proto diagnostic_severity=ERROR diagnostic_summary="Failed to retrieve Incus ImageServer" tf_proto_version=6.6 tf_rpc=ApplyResourceChange timestamp="2024-08-20T20:32:16.070+0900"
2024-08-20T20:32:16.071+0900 [DEBUG] State storage *statemgr.Filesystem declined to persist a state snapshot
2024-08-20T20:32:16.072+0900 [ERROR] vertex "incus_instance.forgejo" error: Failed to retrieve Incus ImageServer
╷
│ Error: Failed to retrieve Incus ImageServer
│
│   with incus_instance.forgejo,
│   on forgejo.tf line 6, in resource "incus_instance" "forgejo":
│    6: resource "incus_instance" "forgejo" {
│
│ The remote isn't a private server
╵
2024-08-20T20:32:16.076+0900 [DEBUG] provider.stdio: received EOF, stopping recv loop: err="rpc error: code = Unavailable desc = error reading from server: EOF"
2024-08-20T20:32:16.077+0900 [DEBUG] provider: plugin process exited: path=.terraform/providers/registry.opentofu.org/lxc/incus/0.1.3/linux_amd64/terraform-provider-incus_v0.1.3 pid=1165472
2024-08-20T20:32:16.077+0900 [DEBUG] provider: plugin exited
```

**Comment by maveonair** [comment]
@benoitjpnet we did some changes in the last 3 weeks to the provider such as update the documentation link.

So in your example, this should work:

```hcl
resource "incus_instance" "forgejo" {
  name  = "forgejo"
  image = "forgejo-8.0.1-1benoitjpnet"

  device {
  [...]
  }
}
```

as I assume that the server on which the image is located is also the one on which you want to deploy it?

---

Also take a look at https://github.com/lxc/terraform-provider-incus/pull/131, where in the future you can use a data source to reference to an existing image. 

```hcl
data "incus_image" "forgejo" {
  name = "forgejo-8.0.1-1benoitjpnet"
}

output "fingerprint" {
  value = data.incus_image.my_image.fingerprint
}

resource "incus_instance" "forgejo" {
  name  = "forgejo"
  image = incus_image.forgejo.fingerprint

  device {
  [...]
  }
}
```

**Comment by stgraber** [comment]
Marking as Incomplete as we wait for @benoitjpnet to report back about current version of the provider.

**Comment by benoitjpnet** [comment]
>    as I assume that the server on which the image is located is also the one on which you want to deploy it?

No, it is not.

I updated to 0.1.4 and still have the error.

```shell
benpro on chiisai in ~/Devel/incus via 💠 default took 11s569ms 
at 05:25:11 ❯ tofu apply
incus_storage_pool.default: Refreshing state... [name=default]
incus_storage_volume.forgejo_var_lib_forgejo: Refreshing state... [name=forgejo_var_lib_forgejo]
incus_storage_volume.laminar_var_lib_incus: Refreshing state... [name=laminar_var_lib_incus]
incus_storage_volume.laminar_var_lib_laminar: Refreshing state... [name=laminar_var_lib_laminar]
incus_instance.laminar: Refreshing state... [name=laminar]

OpenTofu used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

OpenTofu will perform the following actions:

  # incus_instance.forgejo will be created
  + resource "incus_instance" "forgejo" {
      + config           = {}
      + ephemeral        = false
      + image            = "laminar.incus:forgejo-8.0.1-1benoitjpnet"
      + ipv4_address     = (known after apply)
      + ipv6_address     = (known after apply)
      + limits           = {}
      + mac_address      = (known after apply)
      + name             = "forgejo"
      + profiles         = [
          + "default",
        ]
      + running          = true
      + status           = (known after apply)
      + target           = (known after apply)
      + type             = "container"
      + wait_for_network = true

      + device {
          + name       = "var_lib_forgejo"
          + properties = {
              + "path"   = "/var/lib/forgejo"
              + "pool"   = "default"
              + "source" = "forgejo_var_lib_forgejo"
            }
          + type       = "disk"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  OpenTofu will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

incus_instance.forgejo: Creating...
╷
│ Error: Failed to retrieve Incus ImageServer
│ 
│   with incus_instance.forgejo,
│   on forgejo.tf line 6, in resource "incus_instance" "forgejo":
│    6: resource "incus_instance" "forgejo" {
│ 
│ The remote isn't a private server
╵

benpro on chiisai in ~/Devel/incus via 💠 default took 2s264ms 
at 05:25:14 ❌1 ❯ incus remote ls
+---------------------------+------------------------------------+---------------+-------------+--------+--------+--------+
|           NAME            |                URL                 |   PROTOCOL    |  AUTH TYPE  | PUBLIC | STATIC | GLOBAL |
+---------------------------+------------------------------------+---------------+-------------+--------+--------+--------+
| docker                    | https://registry-1.docker.io/v2    | oci           | none        | YES    | NO     | NO     |
+---------------------------+------------------------------------+---------------+-------------+--------+--------+--------+
| images                    | https://images.linuxcontainers.org | simplestreams | none        | YES    | NO     | NO     |
+---------------------------+------------------------------------+---------------+-------------+--------+--------+--------+
| incus.home.arpa (current) | https://incus.home.arpa:8443       | incus         | tls         | NO     | NO     | NO     |
+---------------------------+------------------------------------+---------------+-------------+--------+--------+--------+
| laminar.incus             | https://laminar.incus:8443         | incus         | tls         | YES    | NO     | NO     |
+---------------------------+------------------------------------+---------------+-------------+--------+--------+--------+
| local                     | unix://                            | incus         | file access | NO     | YES    | NO     |
+---------------------------+------------------------------------+---------------+-------------+--------+--------+--------+

```

**Comment by benoitjpnet** [comment]
#131 is also failing:

```shell
benpro on chiisai in ~/Devel/incus via 💠 default 
at 05:38:00 ❯ cat ~/.tofurc
provider_installation {
  dev_overrides {
    "lxc/incus" = "/home/benpro/Devel/terraform-provider-incus"
  }
}
```

```hcl
resource "incus_storage_volume" "forgejo_var_lib_forgejo" {
  name = "forgejo_var_lib_forgejo"
  pool = incus_storage_pool.default.name
}

data "incus_image" "forgejo" {
  name = "forgejo-8.0.1-1benoitjpnet"
  remote = "laminar.incus"
}

resource "incus_instance" "forgejo" {
  name  = "forgejo"
  #image = "laminar.incus:forgejo-8.0.1-1benoitjpnet"
  image = "data.incus_image.forgejo.fingerprint"

  device {
    name = "var_lib_forgejo"
    type = "disk"
    properties = {
      path   = "/var/lib/forgejo"
      source = incus_storage_volume.forgejo_var_lib_forgejo.name
      pool   = incus_storage_pool.default.name
    }
  }
}
```

```
tofu apply
╷
│ Warning: Provider development overrides are in effect
│ 
│ The following provider development overrides are set in the CLI configuration:
│  - lxc/incus in /home/benpro/Devel/terraform-provider-incus
│ 
│ The behavior may therefore not match any released version of the provider and applying changes may cause the state to become incompatible with published releases.
╵
incus_storage_pool.default: Refreshing state... [name=default]
data.incus_image.forgejo: Reading...
incus_storage_volume.forgejo_var_lib_forgejo: Refreshing state... [name=forgejo_var_lib_forgejo]
incus_storage_volume.laminar_var_lib_incus: Refreshing state... [name=laminar_var_lib_incus]
incus_storage_volume.laminar_var_lib_laminar: Refreshing state... [name=laminar_var_lib_laminar]
incus_instance.laminar: Refreshing state... [name=laminar]

OpenTofu used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

OpenTofu planned the following actions, but then encountered a problem:

  # incus_instance.forgejo will be created
  + resource "incus_instance" "forgejo" {
      + config           = {}
      + ephemeral        = false
      + image            = "data.incus_image.forgejo.fingerprint"
      + ipv4_address     = (known after apply)
      + ipv6_address     = (known after apply)
      + limits           = {}
      + mac_address      = (known after apply)
      + name             = "forgejo"
      + profiles         = [
          + "default",
        ]
      + running          = true
      + status           = (known after apply)
      + target           = (known after apply)
      + type             = "container"
      + wait_for_network = true

      + device {
          + name       = "var_lib_forgejo"
          + properties = {
              + "path"   = "/var/lib/forgejo"
              + "pool"   = "default"
              + "source" = "forgejo_var_lib_forgejo"
            }
          + type       = "disk"
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
╷
│ Error: Failed to retrieve Incus ImageServer
│ 
│   with data.incus_image.forgejo,
│   on forgejo.tf line 6, in data "incus_image" "forgejo":
│    6: data "incus_image" "forgejo" {
│ 
│ The remote isn't a private server

```

**Comment by maveonair** [comment]
@stgraber I did some troubleshooting and I think the problem is that the public server with the public image is set to the protocol “Incus” and therefore the Terraform provider consider it as an instance server and not an image server: https://github.com/lxc/terraform-provider-incus/blob/main/internal/provider-config/config.go#L170-L192 and then we hit the following Code in Incus itself: https://github.com/lxc/incus/blob/cbebce1dcd5f15887967058c8f6fec27cf0da2a2/shared/cliconfig/remote.go#L73

In other words, it looks like this is working as intended? So to use the public server, should the procotol be set to `simplestreams` instead?

**Comment by benoitjpnet** [comment]
Is this really expected?

What I want to use is not simple streams but public server.

Documentation excerpt:

    Public Incus servers
    Incus servers that are used solely to serve images and do not run     
instances themselves.

    To make an Incus server publicly available over the network on   port 
8443, set the core.https_address configuration option to :8443  and do 
not configure any authentication methods (see How to expose Incus to the 
network for more information). Then set the images that you want to share 
to public.

Sep 28, 2024 12:49:28 AM Fabian Mettler ***@***.***>:

**Comment by benoitjpnet** [comment]
I don’t think the bug report should be flagged as incomplete. Let me know how I can help.

**Comment by maveonair** [comment]
It took me more time to think about it than to make the necessary code changes 😄 . But the following PR will fix this bug: https://github.com/lxc/terraform-provider-incus/pull/150
